### PR TITLE
feat: add filter module

### PR DESCRIPTION
**Pull request by davidsoderberg**

### CODE REVIEW

File: apps/api/src/app/events/e2e/trigger-event.e2e.ts [hunk 1]
```diff
@@ -524,6 +524,72 @@ describe('Trigger event - /v1/events/trigger (POST)', function () {
     const subscriberIds = messages.map((message) => message._subscriberId).filter(isUnique);
     expect(subscriberIds.length).to.equal(4);
   });
+
+  it('should not filter a message with correct payload', async function () {
```
Comment (scopsy) [inline apps/api/src/app/events/e2e/trigger-event.e2e.ts:528]: Maybe worth adding a couple of other steps here to make sure only the intended runs? Or to do a negated value to make sure it's not running? This test might pass even if the filter is not working or implemented. wdyt?

File: apps/api/src/app/events/usecases/send-message/send-message.usecase.ts [hunk 2]
```diff
@@ -35,4 +44,19 @@ export class SendMessage {
...
   }
+
+  private async filter(command: SendMessageCommand) {
+    const data = await this.getFilterData(command);
+
+    return matchMessageWithFilters(command.step, data);
+  }
+
+  private async getFilterData(command: SendMessageCommand) {
+    const subscriber = await this.subscriberRepository.findById(command.subscriberId);
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/send-message/send-message.usecase.ts:55]: Is the findById looks by subscriberId? Or is it _subscriberId. Also is this query is mandatory always or only when we have the `subscriber` filters? Can we only fetch it when needed?

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
-  return steps.filter((message) => {
-    if (message.filters?.length) {
-      const foundFilter = message.filters.find((filter) => {
-        if (filter.type === 'GROUP') {
-          return handleGroupFilters(filter, payloadVariables);
-        }
-
-        return false;
-      });
+export function matchMessageWithFilters(step: NotificationStepEntity, payloadVariables: any): boolean {
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:4]:
```suggestion
export function matchMessageWithFilters(step: NotificationStepEntity, payloadVariables: any): boolean {
```

Can we add typing for the payload? The same type used by the API or Novu SDK

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts [hunk 4]
```diff
@@ -1,24 +1,23 @@
...
+  if (!Array.isArray(step?.filters)) {
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:5]:
```suggestion
  if (step?.filters || !Array.isArray(step?.filters)) {
```

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts [hunk 5]
```diff
@@ -1,24 +1,23 @@
...
+    return true;
+  }
+  const firstFilter = step.filters[0]?.children;
+  if (!firstFilter || (Array.isArray(firstFilter) && firstFilter.length === 0)) {
+    return true;
+  }
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:11]: Could you eloborate here? Not sure I understand this part. Or can we make it so it works without accessing the '0' position ?

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts [hunk 6]
```diff
@@ -44,12 +43,47 @@ function handleOrFilters(filter, payloadVariables) {
 }
 
 function processFilterEquality(i, payloadVariables) {
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:45]: Could we please add typings here?

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts [hunk 7]
```diff
@@ -44,12 +43,47 @@ function handleOrFilters(filter, payloadVariables) {
...
+    return !payloadVariable.includes(value);
+  }
+  if (i.operator === 'IN') {
+    return payloadVariable.includes(value);
   }
 
   return false;
 }
+
+function parseValue(originValue, parsingValue) {
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:76]: Typings here would be great too 🙏

File: apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx [hunk 8]
```diff
@@ -121,6 +125,24 @@ const WorkflowEditorPage = ({
             addStep={addStep}
             setSelectedNodeId={setSelectedNodeId}
           />
+          <When truthy={![StepTypeEnum.DIGEST, null].includes(selectedChannel)}>
```
Comment (scopsy) [inline apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx:128]: Could we extract this to a variable or something? Hard to follow this statement.

File: apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx [hunk 9]
```diff
@@ -153,13 +175,50 @@ const WorkflowEditorPage = ({
...
+                  <NavSection>
+                    {steps.map((i, index) => {
+                      if (index !== activeStep) {
+                        return null;
+                      }
+                      if (!i.filters || !Array.isArray(i.filters)) {
+                        return null;
+                      }
+
+                      return i.filters.map((group) => {
```
Comment (scopsy) [inline apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx:204]: Can we maybe also add some component tests here to make sure we cover all the cases?

**Review by scopsy — COMMENTED** [review]
Great work @davidsoderberg! Added a couple of comments here 🙏

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts (hunk 5 above)
Comment (davidsoderberg) [reply apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:11]: That filters is an array is because we like to select multiple filter groups in the future like 'and' and 'or' 😄

Comment (davidsoderberg) [reply apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:11]: Oh wait I will refactor this...

File: apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx (hunk 8 above)
Comment (davidsoderberg) [reply apps/web/src/pages/templates/workflow/WorkflowEditorPage.tsx]: I have refactored this now 😄

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts (hunk 4 above)
Comment (davidsoderberg) [reply apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:11]: should it not be `!step?.filters`

File: apps/api/src/app/events/e2e/trigger-event.e2e.ts (hunk 1 above)
Comment (davidsoderberg) [reply apps/api/src/app/events/e2e/trigger-event.e2e.ts:528]: Yeah this was what I thought but I think we had some communication problem in Discord, will fix this 😄

File: apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts (hunk 4 above)
Comment (scopsy) [reply apps/api/src/app/events/usecases/trigger-event/message-filter.matcher.ts:11]: Yes correct, my bad :P
